client.go

```diff
@@ -672,6 +688,39 @@ func (c *Client) Start() (addr net.Addr, err error) {
 	return
 }
 
+// checkProtoVersion verifies that the client is returning a compatible protocol
+// version, and sets the PluginSet that will be used for the negotiated
+// version.
+func (c *Client) checkProtoVersion(protoVersion string) error {
+	serverVersion, err := strconv.Atoi(protoVersion)
+	if err != nil {
+		return fmt.Errorf("Error parsing protocol version %q: %s", protoVersion, err)
+	}
+
+	// record these for the error message
+	var clientVersions []int
+
+	// all versions, including the legacy ProtocolVersion have been added to
+	// the versions set
+	for version := range c.config.VersionedPlugins {
+		clientVersions = append(clientVersions, version)
+
+		if serverVersion != version {
+			continue
+		}
+
+		// set the Plugins value to the compatible set, so the version
+		// doesn't need to be passed through to the ClientProtocol
+		// implementation.
+		c.config.Plugins = c.config.VersionedPlugins[version]
```

Comment:
Could we store the selected version in the client for later retrieval? This would be helpful so we can show a deprecation notice if the server is running an out of date version.

Reply:
Good idea. I added a `NegotiatedVersion` field to the client.